xarray/core/dataset.py

```diff
@@ -1433,11 +1449,68 @@ def __setitem__(self, key: Hashable, value) -> None:
         variable.
         """
         if utils.is_dict_like(key):
-            raise NotImplementedError(
-                "cannot yet use a dictionary as a key to set Dataset values"
-            )
+            from .dataarray import DataArray
 
-        self.update({key: value})
+            # check for consistency first
+            if isinstance(value, Dataset):
+                missing_vars = [
+                    str(name) for name in value.data_vars if name not in self.data_vars
+                ]
+                if len(missing_vars) > 0:
+                    raise KeyError(
+                        "Variables {} in new values not available in dataset!".format(
+                            ",".join(missing_vars)
+                        )
+                    )
```

Comment:
nit:
```suggestion
                missing_vars = [
                    name for name in value.data_vars if name not in self.data_vars
                ]
                if missing_vars:
                    raise KeyError(
                        f"Variables {missing_vars} in new values not available in dataset:\n{dataset}"
                        )
                    )
```

- uses "implicit truthiness" for Python lists, which is more canonical Python style
- no need to build the string of variables manually when a list has a fine repr already
- f-strings are a little more readable than `.format()`
- prints the existing dataset, for clarity about what went wrong

Reply:
good points, thanks

Reply:
unfortunately, KeyError does not display line breaks "\n".  So I changed it to ValueError. Or do you have a better suggestion?

Reply:
@matzegoebel thanks for making these changes. Are you sure there's a difference between how `KeyError` and `ValueError` display the error message? That would be surprising (and interesting!).

Reply:
At least it does for me. I found this: https://stackoverflow.com/questions/46892261/new-line-on-error-message-in-keyerror-python-3-3

Reply:
But yeah it's quite weird

Reply:
TIL! Thanks!

It has surprisingly bad effects — this line is 696 characters long!

```python
In [5]: raise KeyError(f"{ds}")
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-5-cc3fff066a44> in <module>
----> 1 raise KeyError(f"{ds}")

KeyError: '<xarray.Dataset>\nDimensions:  (lat: 25, lon: 53, time: 2920)\nCoordinates:\n  * lat      (lat) float32 75.0 72.5 70.0 67.5 65.0 ... 25.0 22.5 20.0 17.5 15.0\n  * lon      (lon) float32 200.0 202.5 205.0 207.5 ... 322.5 325.0 327.5 330.0\n  * time     (time) datetime64[ns] 2013-01-01 ... 2014-12-31T18:00:00\nData variables:\n    air      (time, lat, lon) float32 241.2 242.5 243.5 ... 296.5 296.2 295.7\nAttributes:\n    Conventions:  COARDS\n    title:        4x daily NMC reanalysis (1948)\n    description:  Data is from NMC initialized reanalysis\\n(4x/day).  These a...\n    platform:     Model\n    references:   http://www.esrl.noaa.gov/psd/data/gridded/data.ncep.reanaly...'
```